scm-step/src/main/java/org/jenkinsci/plugins/workflow/steps/scm/GitStep.java

```diff
@@ -78,6 +119,58 @@ public DescriptorImpl() {
             GitSCM.class.hashCode();
         }
 
+        // copy/paste from GitSCM.DescriptorImpl
+
+        public ListBoxModel doFillCredentialsIdItems(@AncestorInPath Item project,
+                                                     @QueryParameter String url) {
+            if (project == null || !project.hasPermission(Item.CONFIGURE)) {
+                return new StandardListBoxModel();
+            }
+            return new StandardListBoxModel()
+                    .withEmptySelection()
+                    .withMatching(
+                            GitClient.CREDENTIALS_MATCHER,
+                            CredentialsProvider.lookupCredentials(StandardCredentials.class,
+                                    project,
+                                    ACL.SYSTEM,
+                                    GitURIRequirementsBuilder.fromUri(url).build())
+                    );
+        }
+
+        public FormValidation doCheckUrl(@AncestorInPath Item project,
+                                         @QueryParameter String credentialsId,
+                                         @QueryParameter String value) throws IOException, InterruptedException {
+
+            if (project == null || !project.hasPermission(Item.CONFIGURE)) {
+                return FormValidation.ok();
+            }
+
+            String url = Util.fixEmptyAndTrim(value);
+            if (url == null)
+                return FormValidation.error("Please enter Git repository.");
+
+            if (url.indexOf('$') >= 0)
+                // set by variable, can't validate
+                return FormValidation.ok();
+
+            // get git executable on master
+            final EnvVars environment = new EnvVars(EnvVars.masterEnvVars);
+
+            GitClient git = Git.with(TaskListener.NULL, environment)
+                    .using(GitTool.getDefaultInstallation().getGitExe())
+                    .getClient();
+            git.addDefaultCredentials(lookupCredentials(project, credentialsId, url));
```

Comment:
Do not pass in a null `credentialsId` and then pass null to `addDefaultCredentials`. If `credentialsId != null`, then verify that `lookupCredentials` returns non-null; otherwise skip the call to `addDefaultCredentials`.